feat: Add flagz endpoint for kubelet

zhifei92 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Adds a /flagz endpoint for kubelet
part of https://github.com/kubernetes/enhancements/issues/4828
#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Added a /flagz endpoint for kubelet endpoint
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
KEP : https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/4828-component-flagz/README.md
<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
kubectl get --raw "/api/v1/nodes/{nodename}/proxy/flagz"

```docs
kubelet flags
Warning: This endpoint is not meant to be machine parseable, has no formatting compatibility guarantees and is for debugging purposes only.

address=0.0.0.0
allowed-unsafe-sysctls=[]
anonymous-auth=true
application-metrics-count-limit=100
authentication-token-webhook=false
authentication-token-webhook-cache-ttl=2m0s
authorization-mode=AlwaysAllow
authorization-webhook-cache-authorized-ttl=5m0s
authorization-webhook-cache-unauthorized-ttl=30s
boot-id-file=/proc/sys/kernel/random/boot_id
bootstrap-kubeconfig=/etc/kubernetes/bootstrap-kubelet.conf
cert-dir=/var/lib/kubelet/pki
cgroup-driver=cgroupfs
cgroup-root=
cgroups-per-qos=true
client-ca-file=
cloud-config=
cloud-provider=
cluster-dns=[]
cluster-domain=
config=/var/lib/kubelet/config.yaml
config-dir=
container-hints=/etc/cadvisor/container_hints.json
container-log-max-files=5
container-log-max-size=10Mi
container-runtime-endpoint=unix:///var/run/containerd/containerd.sock
containerd=/run/containerd/containerd.sock
containerd-namespace=k8s.io
contention-profiling=false
cpu-cfs-quota=true
cpu-cfs-quota-period=100ms
cpu-manager-policy=none
cpu-manager-policy-options=
cpu-manager-reconcile-period=10s
enable-controller-attach-detach=true
enable-debugging-handlers=true
enable-load-reader=false
enable-server=true
enforce-node-allocatable=[pods]
event-burst=100
event-qps=50
```

zhifei92 commented:
/hold Waiting for PR #128811, which depends on the included `componentKubelet`.

yongruilin commented:
/sig instrumentation
/cc @yongruilin

richabanker commented:
Hello @zhifei92 is this PR still WIP?

zhifei92 commented:
> Hello @zhifei92 is this PR still WIP? 

@richabanker  Hi, it is now ready for review.

richabanker commented on cmd/kubelet/app/server.go:
```diff
@@ -266,6 +263,13 @@ is checked every 20 seconds (also configurable with a flag).`,
 				KubeletConfiguration: *kubeletConfig,
 			}
 
+			if utilfeature.DefaultFeatureGate.Enabled(zpagesfeatures.ComponentFlagz) {
```
Does this need to move above the kubeletServer initialization that's happening above?
Also if possible, could you also add a snippet of the response you get from kubelet's /flagz page after this change in the PR description? I was testing the new endpoint in a kind cluster, and added what I was getting back for the apiserver [here](https://github.com/kubernetes/kubernetes/pull/127581)

richabanker replied:
ah I think I was looking at the older version of the code earlier. You now have added `Flagz` as a field in `kubeletServer` rather than `kubeletConfig`

richabanker commented:
/ok-to-test

richabanker commented on pkg/kubelet/kubelet.go:
```diff
@@ -614,6 +616,7 @@ func NewMainKubelet(kubeCfg *kubeletconfiginternal.KubeletConfiguration,
 		nodeStatusMaxImages:            nodeStatusMaxImages,
 		tracer:                         tracer,
 		nodeStartupLatencyTracker:      kubeDeps.NodeStartupLatencyTracker,
+		Flagz:                          flagz,
```
Does this need to be exported?

zhifei92 replied:
> exported

No need for exported; fixed.

richabanker commented on pkg/kubelet/server/server.go:
```diff
@@ -206,7 +215,7 @@ func ListenAndServeKubeletReadOnlyServer(
 	tp oteltrace.TracerProvider) {
 	klog.InfoS("Starting to listen read-only", "address", address, "port", port)
-	s := NewServer(host, resourceAnalyzer, checkers, nil, nil)
+	s := NewServer(host, resourceAnalyzer, checkers, nil, nil, nil)
```
We should have consistency with this and `ListenAndServeKubeletServer` otherwise the 2 kubelet processes running on different ports will show different behaviors

zhifei92 replied:
Fixed

zhifei92 commented on staging/src/k8s.io/kubelet/config/v1beta1/types.go:
```diff
@@ -89,6 +90,8 @@ const (
 	metav1.TypeMeta `json:",inline"`
 
+	// Flagz is the Reader interface to get flags for flagz page.
+	Flagz flagz.Reader
```
Fixed, thx.

richabanker commented:
/lgtm
but would defer to sig-node to verify that the right flags are instrumented here.

tallclair commented:
/triage accepted
/assign

tallclair commented on cmd/kubelet/app/server.go:
```diff
@@ -1338,7 +1342,8 @@ func createAndInitKubelet(kubeServer *options.KubeletServer,
 		kubeServer.NodeStatusMaxImages,
-		kubeServer.KubeletFlags.SeccompDefault || kubeServer.KubeletConfiguration.SeccompDefault)
+		kubeServer.KubeletFlags.SeccompDefault || kubeServer.KubeletConfiguration.SeccompDefault,
+		kubeServer.Flagz)
```
nit: Can we add `Flagz` to the kubedeps (kubelet.Dependencies) instead?

zhifei92 replied:
I fully agree with your suggestion

tallclair reviewed (commented):
Similar to the comments on https://github.com/kubernetes/kubernetes/pull/128811#pullrequestreview-2486010076, this needs auth configured. I definitely think it makes sense to combine this with the configz permission, since there's a lot of overlap between the flags and config file. I also feel more strongly that we don't expose flags on the readonly port.

/cc @vinayakankugoyal

zhifei92 commented:
> I definitely think it makes sense to combine this with the configz permission, since there's a lot of overlap between the flags and config file.
> 
> /cc @vinayakankugoyal

Therefore, we will combine the configz endpoint and its associated permissions, and introduce statusz as a new subresource with its own authorization rules?

zhifei92 committed:
Add flatz to kubelet auth.

Fix: Move flagz to InstallDebuggingHandlers.

Move flagz  to the lower switch statemen

Fix linter

Drop the function parameter for the read-only server.

zhifei92 commented:
/retest-required

zhifei92 commented:
Friendly ping @tallclair, if you have time, could you please take another look?

tallclair commented on pkg/kubelet/server/auth.go:
```diff
@@ -99,7 +101,8 @@ func (n nodeAuthorizerAttributesGetter) GetRequestAttributes(u user.Info, r *htt
 		case isSubpath(requestPath, healthz.DefaultHealthzPath):
 			subresources = append(subresources, "healthz")
-		case isSubpath(requestPath, configz.DefaultConfigzPath):
+		case isSubpath(requestPath, configz.DefaultConfigzPath),
+			isSubpath(requestPath, flagz.DefaultFlagzPath):
```
move to the lower switch statement, out of the `KubeletFineGrainedAuthz` block

zhifei92 replied:
Done, thx.

zhifei92 replied:
> move to the lower switch statement, out of the `KubeletFineGrainedAuthz` block

@tallclair Currently, `configz` is inside `KubeletFineGrainedAuthz`. Do we need to move `flagz` outside of `KubeletFineGrainedAuthz`?

tallclair replied:
That's fine. The endpoints in this switch statement have the `proxy` subresource appended for backwards compatibility. Since `flagz` is a new endpoint, it doesn't need to support the proxy permission.

tallclair commented on pkg/kubelet/server/auth.go:
```diff
@@ -99,7 +101,8 @@ func (n nodeAuthorizerAttributesGetter) GetRequestAttributes(u user.Info, r *htt
 		case isSubpath(requestPath, healthz.DefaultHealthzPath):
 			subresources = append(subresources, "healthz")
-		case isSubpath(requestPath, configz.DefaultConfigzPath):
+		case isSubpath(requestPath, configz.DefaultConfigzPath),
+			isSubpath(requestPath, flagz.DefaultFlagzPath):
```
IIRC, @vinayakankugoyal wanted a separate authz subresource for this? I'm OK with either way. There's a lot of overlap between config file options & flags, so it makes sense to combine these from a security perspective. Not sure which is more expected from a UX perspective.

tallclair commented on pkg/kubelet/server/server.go:
```diff
@@ -413,6 +419,13 @@ func (s *Server) InstallDefaultHandlers() {
+			s.addMetricsBucketMatcher("flagz")
+			flagz.Install(s.restfulCont, ComponentKubelet, s.flagz)
+		}
+	}
```
Move this to `InstallDebugHandlers`

zhifei92 replied:
Done

zhifei92 commented:
/retest-required

zhifei92 commented:
/retest-required

zhifei92 commented:
@tallclair @vinayakankugoyal  Are there any other issues?

tallclair commented on pkg/kubelet/server/server.go:
```diff
@@ -210,11 +213,12 @@ func ListenAndServeKubeletReadOnlyServer(
 	tp oteltrace.TracerProvider) {
 	klog.InfoS("Starting to listen read-only", "address", address, "port", port)
-	s := NewServer(host, resourceAnalyzer, checkers, nil, nil)
+	s := NewServer(host, resourceAnalyzer, checkers, flagz, nil, nil)
```
nit: you can just pass nil here and drop the function parameter since it's never served by the readonly server.

tallclair reviewed (commented):
/lgtm
/approve

zhifei92 commented:
/cc  @haircommander @SergeyKanzhelev Please take a look. thx~

richabanker commented:
/approve

zhifei92 commented:
@richabanker @tallclair Need lgtm again.

richabanker commented:
/lgtm

zhifei92 commented:
/unhold

zhifei92 commented:
/test pull-kubernetes-e2e-kind-ipv6
